packages-exp/auth-exp/src/core/auth/auth_impl.ts

```diff
@@ -0,0 +1,124 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { getApp } from '@firebase/app-exp';
+import { FirebaseApp } from '@firebase/app-types-exp';
+
+import { Auth, Config, Dependencies } from '../../model/auth';
+import { User } from '../../model/user';
+import { AuthErrorCode } from '../errors';
+import { Persistence } from '../persistence';
+import { PersistenceUserManager } from '../persistence/persistence_user_manager';
+import { assert } from '../util/assert';
+import { ClientPlatform, getClientVersion } from '../util/version';
+
+interface AsyncAction {
+  (): Promise<void>;
+}
+
+export const DEFAULT_API_HOST = 'identitytoolkit.googleapis.com';
+export const DEFAULT_API_SCHEME = 'https';
+
+class AuthImpl implements Auth {
+  currentUser: User | null = null;
+  private operations: Promise<void>;
+  private persistenceManager?: PersistenceUserManager;
+
+  constructor(
+    public readonly name: string,
+    public readonly config: Config,
+    persistenceHierarchy: Persistence[]
+  ) {
+    this.operations = Promise.resolve();
+
+    // This promise is intended to float; auth initialization happens in the
+    // background, meanwhile the auth object may be used by the app.
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
```

Comment:
will this still call the promise? is there a way to test that?

Reply:
It will, promises always execute. There's no concept of "hot" and "cold" promises like there are with observables. Since this.operations is initialized as `Promise.resolve()`, and `queue()` just puts the callback in a `.then()`, the initial promise is resolved and so the callback here will immediately execute.

There's not really a way to test that it runs on its own since the test has to `await` _something_, but my tests do cover this code by awaiting a later operation (but note it's _not_ the act of awaiting that triggers this code to run)